evm/combine-signatures.js

```diff
@@ -0,0 +1,273 @@
+
+            validateParameters({
+                isNonEmptyString: { sourceChain, sourceAddress },
+                isValidAddress: { contractAddress },
+                isKeccak256Hash: { commandId, payloadHash, sourceTxHash },
+                isValidNumber: { sourceEventIndex },
+            });
+
+            const chainId = chain.chainId;
+            const commandID = commandId || getRandomBytes32();
+            const command = 'approveContractCall';
+            const params = defaultAbiCoder.encode(
+                ['string', 'string', 'address', 'bytes32', 'bytes32', 'uint256'],
+                [sourceChain, sourceAddress, contractAddress, payloadHash, sourceTxHash, sourceEventIndex],
+            );
+
+            const data = defaultAbiCoder.encode(
+                ['uint256', 'bytes32[]', 'string[]', 'bytes[]'],
+                [chainId, [commandID], [command], [params]],
+            );
+
+            const dataHash = hashMessage(arrayify(keccak256(data)));
+
+            printInfo('Original bytes message (pre-hash)', data);
+            printInfo('Message hash for validators to sign', dataHash);
+
+            break;
+        }
+
+        case 'constructBatch': {
+            const { message } = options;
+
+            validateParameters({ isValidCalldata: { message } });
+
+            const {
+                addresses: validatorAddresses,
+                weights,
+                threshold,
+                keyID: expectedKeyId,
+            } = await getEVMAddresses(config, chain.id, options);
+
+            const validatorWeights = {};
+
+            validatorAddresses.forEach((address, index) => {
+                validatorWeights[address.toLowerCase()] = weights[index];
+            });
+
+            const signatures = readSignatures();
+
+            const sortedSignatures = signatures.sort((a, b) => {
+                const addressA = getAddressFromPublicKey(`0x${a.pub_key}`).toLowerCase();
+                const addressB = getAddressFromPublicKey(`0x${b.pub_key}`).toLowerCase();
+                return addressA.localeCompare(addressB);
+            });
+
+            const batchSignatures = [];
+
+            let totalWeight = 0;
+
+            const expectedMessageHash = hashMessage(arrayify(keccak256(message)));
+
+            for (const signatureJSON of sortedSignatures) {
+                const keyId = signatureJSON.key_id;
+                const msgHash = `0x${signatureJSON.msg_hash}`;
+                const pubKey = `0x${signatureJSON.pub_key}`;
+                const signature = `0x${signatureJSON.signature}`;
+
+                validateParameters({
+                    isNonEmptyString: { keyId },
+                    isKeccak256Hash: { msgHash },
+                    isValidCalldata: { pubKey, signature },
+                });
+
+                if (expectedKeyId !== keyId) {
```

Comment:
current key id might have changed/rotated. instead check if key id in the file is valid, and matches for all signatures